Initial data structures for incremental smt2 solving

thomasspriggs opened the pull request:
Initial data structures for implementation of incremental smt2 solving.
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->
Standard PR checklist: 
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

TGWDB commented on src/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,166 @@
+
+static bool is_valid_smt_identifier(irep_idt identifier)
+{
+  static const std::regex valid{R"(^[^\|\\]*$)"};
+  return std::regex_match(id2string(identifier), valid);
+}
+
+smt_identifier_termt::smt_identifier_termt(irep_idt identifier, smt_sortt sort)
+  : smt_termt(ID_smt_identifier_term, std::move(sort))
+{
+  INVARIANT(
+    is_valid_smt_identifier(identifier),
+    R"(Identifiers may not contain | or \ characters.)");
```
I'm not sure that this exactly matches the SMT2 specification (I'm looking at pages 95-96), is this meant to match the specification or is the cbmc version slightly different here?

thomasspriggs replied:
My intention here is to check that our internal representation is not escaped and then to perform escaping as part of the printing. I was concerned we could accidentally end up with doubly escaped identifiers. We should probably discus outside of github what the precise requirements should be here.

TGWDB commented on unit/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,89 @@
+TEST_CASE("smt_not_termt upcasts.", "[core][smt2_incremental]")
+{
+  smt_termt not_term{smt_not_termt{smt_bool_literal_termt{false}}};
+  not_term = smt_not_termt{smt_bool_literal_termt{false}};
```
false -> true? for one of the cases

thomasspriggs replied:
Done.

TGWDB commented on unit/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,89 @@
+
+TEST_CASE("smt_identifier_termt upcasts.", "[core][smt2_incremental]")
+{
+  smt_termt term{smt_identifier_termt{"foo", smt_bool_sortt{}}};
+  term = smt_identifier_termt{"bar", smt_bool_sortt{}};
+}
```
Test case for failing the regex?

thomasspriggs replied:
New / separate constructor tests added.

TGWDB replied:
Seems to only test for `|`?

thomasspriggs replied:
I have added a test for `\`.

TGWDB commented on unit/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,89 @@
+  CHECK(bit_vector.get_sort() == smt_bit_vector_sortt{8});
+}
+
+TEST_CASE("smt_termt equality.", "[core][smt2_incremental]")
+{
+  smt_termt false_term = smt_bool_literal_termt{false};
+  CHECK(false_term == false_term);
+  CHECK(false_term == smt_bool_literal_termt{false});
+  smt_termt true_term = smt_bool_literal_termt{true};
+  CHECK(false_term != true_term);
+  CHECK(
+    smt_bit_vector_constant_termt{42, 8} !=
+    smt_bit_vector_constant_termt{12, 8});
+}
```
Equality test using `not_term`?

thomasspriggs replied:
Done.

TGWDB reviewed (commented):
Also no unit tests for the smt_commands?

thomasspriggs commented:
> Also no unit tests for the smt_commands?

Its on the TODO list in the PR description.

TGWDB commented:
For bookkeeping, note that this is part of the SMT backend work with broad explanation and RFC in #6134

martin-cs commented on src/solvers/smt2_incremental/smt_sorts.h:
```diff
@@ -0,0 +1,91 @@
+
+class smt_sort_const_downcast_visitort;
+
+class smt_sortt : protected irept
```
Why define these separately when you could use `typet`?  The type system is essentially identical.  I can see a sort of value in having a `class smt_sortt : public typet` and a `smt_sortt to_smt_sort(typet &t)` function so you can use the C++ type safety to enforce conversion but this really seems like overkill.

martin-cs commented on src/solvers/smt2_incremental/smt_commands.h:
```diff
@@ -0,0 +1,92 @@
+#include <solvers/smt2_incremental/smt_terms.h>
+#include <util/irep.h>
+
+class smt_commandt : protected irept
```
As a sanity test for this, would it be worth integrating these into `smt2_solver` / `smt2_parser`? They should make the separation between parsing and solving much clearer.

thomasspriggs replied:
Yes. Using the new structures for the solving will help sanity test them. I have already found several minor issues by working on solving/converting to string locally. However I feel it is better to get a first version merged and then refine the structures as we work on the solving, rather than perfect the first version before merging the first PR. Integrating the new structures into the parsing would be a nice-to-have. But we are currently thinking of working on exprt -> new structures and new structures -> string before parsing.

martin-cs commented on src/solvers/smt2_incremental/smt_sorts.h:
```diff
@@ -0,0 +1,91 @@
+
+  /// \brief Class for adding the ability to up and down cast smt_sortt to and
+  ///   from irept. These casts are required by other irept derived classes in
+  ///   order to store instances of smt_sortt inside them.
```
Why not just publicly inherit from `irept`?

thomasspriggs replied:
Because I want to restrict the ability to cast to and from `irept`, to the inside of the storage classes. I see the fact that each of these inherit from `irept` to be an implementation detail. Ideally, there should be no need to upcast to `irept` as part of the algorithms which live outside the storage classes. Keeping variables declared as `smt_termt` or `smt_sortt` rather than `irept` where possible will give us more specific compile time type information.

tautschnig replied:
Wouldn't your rationale then suggest using an `irept` member instead of inheriting?

thomasspriggs replied:
> Wouldn't your rationale then suggest using an `irept` member instead of inheriting?

Yes, I could have done that. I could also have implemented this without using `irept` at all and handled lifetime using `std::unique_ptr` or some such. By using `irept` with inheritance this code retains a greater semblance to other parts of the code base. This should aid cbmc developers familiar with other parts of the code base to transition to maintaining or improving this part and it should aid with reusing the existing `pretty`, `operator==` and hashing algorithms defined for `irept`. Otherwise we would need to write more new code to define these pieces of functionality for the new data structures.

martin-cs commented on src/solvers/smt2_incremental/smt_sorts.def:
```diff
@@ -0,0 +1,2 @@
+SORT_ID(bool)
```
I honestly don't see the value of using this over `ID_bool` used in `bool_typet`.

thomasspriggs replied:
Please bear in mind that this file is not only being used to define the IDs of the `smt_sortt` classes. It is also being used to automate writing some of the more repetitive parts of the visitor pattern. :thinking: I should probably add some doxygen to this file explaining how it is used. What would be the value in rationing the number of different IDs used rather than keeping them separate?

martin-cs commented on src/solvers/smt2_incremental/smt_terms.h:
```diff
@@ -0,0 +1,126 @@
+class smt_sortt;
+class smt_term_const_downcast_visitort;
+
+class smt_termt : protected irept, private smt_sortt::storert<smt_termt>
```
Why not just user `exprt` or if you want to do some kind of conversion, then have a simple wrapper around `exprt` that marks the converted values?

thomasspriggs replied:
Because this gives us more compile-time checking of the soundness of the conversion and better guarantees about the separation of concerns. This applies both in terms of the conversion from `exprt` to `smt_termt` and the conversion from `smt_termt` to string or other representation understood by the solver. We are only currently planning a conversion to SMT2 strings, but by keeping the conversion to strings separate it would theoretically allow for more easily adding an additional conversion algorithm such as to `smt-switch` data structures to be added on top as an additional run time option. Without the new data structures, the conversion from `exprt` to `smt_termt` could "leak" unconverted `exprt` classes which can only be caught at run time. The way the visitor is defined will also mean that missing conversions from `smt_termt` to string would also cause a compile error. There has to be at least a placeholder method in place with the `UNIMPLEMENTED` macro. 

Defining the data structure will give us a clean separation between the conversion from `exprt` to `smt_termt` and `smt_termt` to string. It will allow for separately unit testing the two conversions. A new data structure only for use with `smt2` has the same advantages that having separate data structures from the language front ends such as for `Java` and `C` vs the internal `goto_program` representation would have.

martin-cs replied:
I think you can achieve much of the compile time checking at a much lower cost of having something like...

```
class smt_sortt : public typet {};
class smt_termt : public exprtt {};

class smt_translator {
public :
  virtual smt_sortt type_to_sort(const typet &t) = 0;
  virtual smt_termt expr_to_term(const exprt &e) = 0;
}
```
If this is the only way of producing these two types and the rest of the code is `const`-clean and the critical functions take `smt_sortt` and `smt_termt` then it will be very rare to have failed-to-convert bug.

I would suggest an object because these transformations tend to wind up being stateful no matter how good everyone's intent is.  They will likely also need to be `stackable` in some sense.  As soon as you have a couple of options on how to translate each theory / type (IIRC this was on your list of goals?) then you will have a combinatorial explosion of normal forms, so it is not so easy to use types* to enforce normal forms.

Using the visitor to do the actual conversion or caching converted things is smart idea and it is good to see it early integrated into the design here.

*.  I could be wrong, if I am I would love to know because the handling of goto-programs has exactly this issue.

thomasspriggs replied:
I am not in favour of publicly inheriting from `typet`, for the following reasons -

- It would support down casting from `typet` to `smt_sortt` and require checks to be written for these casts.
- It would add public methods to `smt_sortt` such as `subtype()` and `source_location()` which don't necessarily make sense in the context of a sort.
- It would add public access to member functions from `irept`, which would allow direct manipulation of the `irept` data structures through weakly typed member functions such as `move_to_sub`.
- It would add public access to `irept`'s `is_nil` member function, whereas I would prefer to make the classes in this hierarchy non-nullable/nillable. That way `optionalt` can be added on top of these classes only in the places where nullability is a desirable property.

On the point of the translator - Yes, we will need functionality to convert from `exprt` to `termt`. The choice of whether to use member functions or free functions with configuration/state struct parameters to do the conversions is mostly a stylistic choice in my opinion, rather than a crucial difference. This is unless we really need virtual inheritance and multiple different translators. I have found free functions to be a little easier to unit test in some circumstances, which is a point in their favour. I do think it would improve readability to store configuration and state separately assuming we end up needing both though. That way the configuration can be kept const and the state mutable.

martin-cs commented on src/solvers/smt2_incremental/smt_terms.h:
```diff
@@ -0,0 +1,126 @@
+
+  const smt_sortt &get_sort() const;
+
+  void accept(smt_term_const_downcast_visitort &) const;
```
It is good to see visitors!

martin-cs reviewed (commented):
I feel like I must be missing / not understanding a requirement here.  The commands; yes, they are good and would be useful.  But the rest -- why not just use `exprt` and `typet`?  The underlying representation is the same and there are already a load of utility functions.  Also, in the vast majority of cases, the semantics are the same as well.  I can understand that it might be nice to use the C++ type system to mark which are values used in this subsystem or having had some conversions / checks applied but that can be done in a much more lightweight fashion.

SaswatPadhi commented on src/solvers/smt2_incremental/smt_commands.def:
```diff
@@ -0,0 +1,8 @@
+COMMAND_ID(exit)
+COMMAND_ID(get_value)
+COMMAND_ID(set_logic)
+COMMAND_ID(set_option)
```
Are `(push)` and `(pop)` out of scope for this PR?

thomasspriggs replied:
I had not considered them yet. However I think it will make sense to add them early on. So I will likely add them before I move this PR out of draft status.

martin-cs replied:
Would it be possible to add `check-sat-assuming` as well please?

thomasspriggs replied:
Incremental solving should be possible without `check-sat-assuming` right? I'd prefer not to add too many different things up front, so that we can get a MVP working initially and then add incremental improvements to the incremental solving support. There is no reason that more data structures cannot be added as they are needed.

thomasspriggs replied:
I have now added `(push n)` and `(pop n)`.

NlightNFotis commented on unit/solvers/smt2_incremental/smt_sorts.cpp:
```diff
@@ -0,0 +1,32 @@
+{
+  const smt_bit_vector_sortt bit_vec_sort{32};
+  smt_sortt sort{bit_vec_sort};
+  sort = bit_vec_sort;
```
What are these two test cases testing for exactly?

thomasspriggs replied:
It tests that the copy constructor of the `smt_sortt` class can be called with a smt_bit_vector_sortt typed parameter and the same for the assignment operator. Effectively testing the upcast which is carried out in these two operations is accessible. I added these two tests because the same test with the `irept` constructor should fail due to the access restrictions. Writing a test that it does not work with `irept`, could be done using some template meta programming to write a "does not compile" function, but probably would not be straight forward code to for us to read or write.

NlightNFotis commented on unit/solvers/smt2_incremental/smt_sorts.cpp:
```diff
@@ -30,3 +30,50 @@ TEST_CASE(
+
+TEST_CASE("Visiting smt_bool_sortt.", "[core][smt2_incremental]")
+{
+  class : public smt_sort_const_downcast_visitort
```
⛏️ Given that the visitor is nearly the same in both of the following test cases, does it make sense for it to be pulled out somewhere?

thomasspriggs replied:
Yes. I usually err on the side of more duplication in unit tests, in order to make it more obvious why the properties the test is testing should hold. The duplication isn't too bad for the sorts at the moment as there are only a couple of cases. But it will be much worse for the terms and commands and it will be worse if we expand this with more sorts. Therefore I'll come up with a way to factor it out for the terms and commands and then back port the solution to the sorts.

NlightNFotis commented on unit/solvers/smt2_incremental/smt_to_string.cpp:
```diff
@@ -0,0 +1,12 @@
+
+TEST_CASE("Test smt_sortt to string conversion", "[core][smt2_incremental]")
+{
+  CHECK(smt_to_string(smt_bool_sortt{}) == "Bool");
```
👍🏻 I really like how clean these are because of the visitor - I feel like it was a good design choice.

NlightNFotis commented on src/solvers/smt2_incremental/smt_sorts.cpp:
```diff
@@ -10,6 +10,16 @@
 
+bool smt_sortt::operator==(const smt_sortt &other) const
+{
+  return irept::operator==(other);
```
Question: since this just calls the base class' method, with no other change to it, does it make sense to kill this method here and just `using irept::operator==;` in the header file?

thomasspriggs replied:
The `irept::operator==` member function takes an `irept` as its parameter / right hand side. So yes we could add `using irept::operator==;`, but the existing tests which use the `smt_sortt::operator==` would not compile because the implicit upcast to `irept` is banned from outside of the data structure's member functions due to the use of protected inheritance. Also I don't think it would make sense to allow comparisons such as `smt_bool_sortt{} == irept{}`. 1) because sorts should not be stored as `irept` outside of the irep based data structures and 2) because if we know that one side is not a sort at compile time and that the other side is a sort, then we know that the result would always be false.

NlightNFotis reviewed (commented):
Some questions from me.

thomasspriggs commented:
Note: Some of the make build jobs are currently failing CI because the makefile miscounts the tests which are using TEMPLATE_TEST_CASE

thomasspriggs committed:
Add foundational smt sorts

thomasspriggs committed:
Add smt sort downcast visitor

thomasspriggs committed:
Add operator== to smt_sortt

thomasspriggs committed:
Add `smt_sortt::storer` as a way add storage of sorts

So that instances of `smt_sortt` can be stored in other `irept` derived
classes without enabling upcasting of `smt_sortt` to `irept` from
outside of the heirarchy of `irept` based classes.

thomasspriggs committed:
Rewrite unit test counting in python

So that it can be debugged and further features added separately from
the `makefile`. Note that the check for `[.]` was already broken in the
makefile because `[.]` is inserted using the XFAIL macro, but the files
are grepped before macro preprocessing.

thomasspriggs committed:
Include `TEMPLATE_TEST_CASE`s in test count

So that we can use catch's `TEMPLATE_TEST_CASE` without the test count
failing when using make to run the tests. When using
`TEMPLATE_TEST_CASE` the number of tests added to catch's count of test
depends on the number of type arguments supplied. Therefore this update
is needed to count the arguments  to get the correct count.

thomasspriggs committed:
Add smt_terms data structures

thomasspriggs committed:
Add controlled upcasting of smt_termt for storage implementions

So that instaces of smt_termt can be stored inside other non-term ireps.

thomasspriggs committed:
Add smt commands

thomasspriggs committed:
Add smt_command downcast visitor

thomasspriggs committed:
Add class diagram for internal SMT AST classes

This class diagram served as the notes for implementing the initial set
of classes. It is committed to the repository for future reference.

TGWDB commented on src/solvers/smt2_incremental/internal_ast.txt:
```diff
@@ -0,0 +1,255 @@
+@startuml
+' This diagram is written using PlantUML. It can be visualised locally or using
+' the PlantUML website - http://www.plantuml.com/plantuml/uml/.
+' On Linux the visualisation tool can be installed using -
+'   `sudo apt-get install plantuml`
+' Running the `plantuml` binary from the directory where this file exists will
+' give the option to view the file.
+
```
This file is great to have (and also to be able to generate). I just have the general concern that this is not generated or linked from the source code? How easy will it be for the two to lose synchronisation?

thomasspriggs replied:
I think there is a doxygen plugin which can generate a graph. However I remember these being less than ideal. I suggest we either keep the plantUML and live with the reality that it will likely get out of date, or I remove it entirely and put any useful information from the graph nodes into the doxygen above the corresponding classes.

thomasspriggs committed:
Add `smt_termt` visitor

Because this will be used to implement term to string/output stream
conversion. Use of a visitor will allow this printing code to be
separated from the data structure.

TGWDB commented on src/solvers/smt2_incremental/smt_logics.cpp:
```diff
@@ -0,0 +1,47 @@
+  ::accept(*this, id(), std::move(visitor));
+}
+
+smt_logic_quantifier_free_bit_vectorst::smt_logic_quantifier_free_bit_vectorst()
+  : smt_logict{ID_smt_logic_quantifier_free_bit_vectors}
```
There is some tension here since the code here (and later also marked) uses `quantifier_free_bit_vectors` but I think the actual logic is likely to be `QF_AUFBV` (based on what is used now). Note that this is not `QF_BV` which corresponds to quantifier free bit vector theory. (`QF_AUFBV` also allows arrays and free sort and function symbols, more details on logics [here](http://smtlib.cs.uiowa.edu/logics.shtml)).

thomasspriggs replied:
Yes. `quantifier_free_bit_vectors` would be just `QF_BV`. `QF_AUFBV` is what we were using in the old back end up until recently. I should fix the full names to match the abbreviations which get written out to the solver.

TGWDB commented on src/solvers/smt2_incremental/smt_logics.def:
```diff
@@ -0,0 +1,9 @@
+///  * The constant `irep_idt`s used to identify each of the logic classes.
+///  * The member functions of the `smt_logic_const_downcast_visitort` class.
+///  * The type of option checks required to implement `smt_logict::accept`.
+LOGIC_ID(quantifier_free_bit_vectors)
```
logic tension

TGWDB commented on src/solvers/smt2_incremental/smt_logics.h:
```diff
@@ -0,0 +1,78 @@
+  return static_cast<const smt_logict &>(irep);
+}
+
+class smt_logic_quantifier_free_bit_vectorst : public smt_logict
+{
+public:
+  smt_logic_quantifier_free_bit_vectorst();
+};
```
logic tension

TGWDB commented on src/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,166 @@
+  smt_bit_vector_sortt sort)
+  : smt_termt{ID_smt_bit_vector_constant_term, std::move(sort)}
+{
+  INVARIANT(
+    value < power(mp_integer{2}, get_sort().bit_width()),
+    "value must fit in number of bits available.");
+  INVARIANT(
+    !value.is_negative(),
+    "Negative numbers are not supported by bit vector constants.");
+  set(ID_value, integer2bvrep(value, get_sort().bit_width()));
```
Some of the current code in `smt2_conv.cpp` handles wrapping and negative values for conversion between bitvectors and integers (see `integer2bvrep` and `bvrep2integer` for examples), are we sure of these invariants here? Alternatively, do we want to make clear that integers must have their conversion to unsigned representations done first before conversion to bitvector?

thomasspriggs replied:
The `INVARIANT` is correct whilst we haven't implemented/tested the signed -> unsigned conversion. If we decide to implement the conversion in this constructor, then the invariant can be removed. 

Of note is that bitvector terms are not inherently signed or unsigned. Therefore if we were to do the conversion in the constructor then we could end up with potentially surprising behaviour where the term is constructed with a negative value, but the value is positive when we get it back out. Therefore I am tempted to say that the term should only handle positive values and that the conversions should be done as part of the other conversion code pieces.

martin-cs replied:
Yeah, it is one of the differences between the CPROVER type system and the SMT-LIB sorts.  CPROVER has three kinds of bit-vector, signed, unsigned and ... general bit-vector, while SMT-LIB only has signed and unsigned operations.  Working with / interpreting everything as hex / binary and just presenting everything as unsigned would be a good solution.

thomasspriggs replied:
Ok. I think this question is sufficiently answered. So I am going to mark it as resolved.

TGWDB commented on src/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,166 @@
+{
+}
+
+mp_integer smt_bit_vector_constant_termt::value() const
+{
+  return bvrep2integer(get(ID_value), get_sort().bit_width(), false);
+}
```
Like mentioned above, the existing code converts based on whether or not the bitvector is signed, we might need this information and conversion?

thomasspriggs replied:
Yes. We will need to keep track of signedness. But signedness is not a property held by the SMT flavour of bit vectors. Therefore the tracking and conversion should be handled separately.

TGWDB commented on unit/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,221 @@
+  term = smt_bool_literal_termt{false};
+}
+
+TEST_CASE("Test smt_termt.pretty is accessible.", "[core][smt2_incremental]")
```
Technically this is testing both that the `pretty` function is accessible and the output. I guess a change to the base `pretty` function could invalidate this test without any changes to the `smt_termt` code.

TGWDB commented on unit/solvers/smt2_incremental/smt_terms.cpp:
```diff
@@ -0,0 +1,221 @@
+  CHECK_THROWS(smt_identifier_termt{"foo\\ bar", smt_bool_sortt{}});
+}
+
+TEST_CASE("smt_identifier_termt upcasts.", "[core][smt2_incremental]")
```
No actual `CHECK` inside this test case, I guess it's a style choice?

thomasspriggs replied:
Yeah, I wrote these upcast tests when I was initially working on the accessibility, at that time I just wanted to be sure that the upcasts compiled. However now that the rest of the tests are in place and full of upcasts, I am not sure that this has value. Therefore I might just remove all the upcast tests.

thomasspriggs replied:
I have now removed these tests, as I was unconvinced of the value they added.

chrisr-diffblue reviewed (approved):
Looks sensible to me.
